### PR TITLE
[EHN] Including FBCNet

### PR DESCRIPTION
**Pull request by bruAristimunha**
This PR is built over the #656.

### CODE REVIEW

**Comment by bruAristimunha** [comment]
@tomMoral.
Small test with BCI competition dataset:

image

**Comment by tomMoral** [comment]
can you rebase/merge main so it is easier to read the diff?

**Comment by bruAristimunha** [comment]
done @tomMoral

File: braindecode/models/fbcnet.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        x = self.spectral_filtering(x)
+
+        x = self.spatial_conv(x)
+        batch_size, channels, _, time = x.shape
+        # Check if time is divisible by stride_factor
+        if time % self.stride_factor != 0:
+            # Pad x to make time divisible by stride_factor
+            padding = self.stride_factor - (time % self.stride_factor)
+            x = torch.nn.functional.pad(x, (0, padding))
+            time += padding  # Update the time dimension after padding
```
Comment (bruAristimunha) [inline braindecode/models/fbcnet.py:192]: Def as layers

File: braindecode/models/modules.py [hunk 2]
```diff
@@ -353,6 +353,47 @@ def _max_norm(self):
...
+    torch.Tensor
+        Output tensor after applying the linear transformation.
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        max_norm: float = 1.0,
+        do_weight_norm: bool = True,
```
Comment (bruAristimunha) [inline braindecode/models/modules.py:382]: No need

**Review by bruAristimunha — COMMENTED** [review]
small comments

**Comment by bruAristimunha** [comment]
To make the revision easy:

The original model code is here:
```python


class FBCNet(nn.Module):
    # just a FBCSP like structure : chan conv and then variance along the time axis
    '''
        FBNet with seperate variance for every 1s. 
        The data input is in a form of batch x 1 x chan x time x filterBand
    '''
    def SCB(self, m, nChan, nBands, doWeightNorm=True, *args, **kwargs):
        '''
        The spatial convolution block
        m : number of sptatial filters.
        nBands: number of bands in the data
        '''
        return nn.Sequential(
                Conv2dWithConstraint(nBands, m*nBands, (nChan, 1), groups= nBands,
                                     max_norm = 2 , doWeightNorm = doWeightNorm,padding = 0),
                nn.BatchNorm2d(m*nBands),
                swish()
                )

    def LastBlock(self, inF, outF, doWeightNorm=True, *args, **kwargs):
        return nn.Sequential(
                LinearWithConstraint(inF, outF, max_norm = 0.5, doWeightNorm = doWeightNorm, *args, **kwargs),
                nn.LogSoftmax(dim = 1))

    def __init__(self, nChan, nTime, nClass = 2, nBands = 9, m = 32,
                 temporalLayer = 'LogVarLayer', strideFactor= 4, doWeightNorm = True, *args, **kwargs):
        super(FBCNet, self).__init__()

        self.nBands = nBands
        self.m = m
        self.strideFactor = strideFactor

        # create all the parrallel SCBc
        self.scb = self.SCB(m, nChan, self.nBands, doWeightNorm = doWeightNorm)
        
        # Formulate the temporal agreegator
        self.temporalLayer = current_module.__dict__[temporalLayer](dim = 3)

        # The final fully connected layer
        self.lastLayer = self.LastBlock(self.m*self.nBands*self.strideFactor, nClass, doWeightNorm = doWeightNorm)

    def forward(self, x):
        x = torch.squeeze(x.permute((0,4,2,3,1)), dim = 4)
        x = self.scb(x)
        x = x.reshape([*x.shape[0:2], self.strideFactor, int(x.shape[3]/self.strideFactor)])
        x = self.temporalLayer(x)
        x = torch.flatten(x, start_dim= 1)
        x = self.lastLayer(x)
        return x
        
class swish(nn.Module):
    '''
    The swish layer: implements the swish activation function
    '''
    def __init__(self):
        super(swish, self).__init__()

    def forward(self, x):
        return x * torch.sigmoid(x)
 
class Conv2dWithConstraint(nn.Conv2d):
    def __init__(self, *args, doWeightNorm = True, max_norm=1, **kwargs):
        self.max_norm = max_norm
        self.doWeightNorm = doWeightNorm
        super(Conv2dWithConstraint, self).__init__(*args, **kwargs)

    def forward(self, x):
        if self.doWeightNorm: 
            self.weight.data = torch.renorm(
                self.weight.data, p=2, dim=0, maxnorm=self.max_norm
            )
        return super(Conv2dWithConstraint, self).forward(x)
    
class LinearWithConstraint(nn.Linear):
    def __init__(self, *args, doWeightNorm = True, max_norm=1, **kwargs):
        self.max_norm = max_norm
        self.doWeightNorm = doWeightNorm
        super(LinearWithConstraint, self).__init__(*args, **kwargs)

    def forward(self, x):
        if self.doWeightNorm: 
            self.weight.data = torch.renorm(
                self.weight.data, p=2, dim=0, maxnorm=self.max_norm
            )
        return super(LinearWithConstraint, self).forward(x)

#%% Support classes for FBNet Implementation
class VarLayer(nn.Module):
    '''
    The variance layer: calculates the variance of the data along given 'dim'
    '''
    def __init__(self, dim):
        super(VarLayer, self).__init__()
        self.dim = dim

    def forward(self, x):
        return x.var(dim = self.dim, keepdim= True)

class StdLayer(nn.Module):
    '''
    The standard deviation layer: calculates the std of the data along given 'dim'
    '''
    def __init__(self, dim):
        super(StdLayer, self).__init__()
        self.dim = dim

    def forward(self, x):
        return x.std(dim = self.dim, keepdim=True)

class LogVarLayer(nn.Module):
    '''
    The log variance layer: calculates the log variance of the data along given 'dim'
    (natural logarithm)
    '''
    def __init__(self, dim):
        super(LogVarLayer, self).__init__()
        self.dim = dim

    def forward(self, x):
        return torch.log(torch.clamp(x.var(dim = self.dim, keepdim= True), 1e-6, 1e6))

class MeanLayer(nn.Module):
    '''
    The mean layer: calculates the mean of the data along given 'dim'
    '''
    def __init__(self, dim):
        super(MeanLayer, self).__init__()
        self.dim = dim

    def forward(self, x):
        return x.mean(dim = self.dim, keepdim=True)

class MaxLayer(nn.Module):
    '''
    The max layer: calculates the max of the data along given 'dim'
    '''
    def __init__(self, dim):
        super(MaxLayer, self).__init__()
        self.dim = dim

    def forward(self, x):
        ma , ima = x.max(dim = self.dim, keepdim=True)
        return ma       
    
```

According to the paper, they did a strange training. They trained for 1500 epochs, with a patience of 200 epochs. Stop and start from the checkpoint for another 600 epochs. They useddddd channel drop data augmentation too. Doing cross-session only in the test session, they reported 76.20 average accuracy, with EEGNetv4 with 73.15. 

I reproduced using moabb with 100 epochs and patience of 25 epochs, using four classes (!). I got the results of:

image

image

Really good results for this neural network. The model is really good with four classes, and with two classes, we have these results:

image

image

I am reporting only the test session.

File: braindecode/models/fbcnet.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    the time axis, inspired by the Filter Bank Common Spatial Pattern (FBCSP)
+    algorithm.
+
+    Notes
+    -----
+    This implementation is not guaranteed to be correct and has not been checked
+    by the original authors; it has only been reimplemented from the paper
+    description and source code [fbcnetcode2021]_. There is a difference in the
+    activation function; in the paper, the ELU is used as the activation function,
+    but in the original code, SiLU is used. We follow the code.
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:48]:
```suggestion
    but in the original code, SiLU is used. We followed the code.
```

File: braindecode/models/fbcnet.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    Parameters
+    ----------
+    n_bands : int or None or List[Tuple[int, int]]], default=9
+        Number of frequency bands. Could
+    n_filters_spat : int, default=32
+        The depth of the depthwise convolutional layer.
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:55]: Unclear

File: braindecode/models/fbcnet.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    n_dim: int, default=3
+        Number of dimensions for the temporal reductor
+    temporal_layer : str, default='LogVarLayer'
+        Type of temporal aggregator layer. Options: 'VarLayer', 'StdLayer',
+        'LogVarLayer', 'MeanLayer', 'MaxLayer'.
+    stride_factor : int, default=4
+        Stride factor for reshaping.
+    activation : nn.Module, default=Swish
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:63]: Selu

File: braindecode/models/fbcnet.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        Activation function class to apply.
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:64]:
```suggestion
        Activation function to apply.
```
In which module?

File: braindecode/models/fbcnet.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    verbose: bool, default False
+        Verbose parameter to create the filter using mne
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:66]: I would remove this parameter, no one will use it.

File: braindecode/models/fbcnet.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        if self.n_times % self.stride_factor != 0:
+            warn(
+                f"Time dimension ({self.n_times}) is not divisible by"
+                f" stride_factor ({self.stride_factor}). Input will be "
+                f"truncated.",
+                UserWarning,
+            )
+
+        # Layers
+        # Following paper nomeclature
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:131]:
```suggestion
        # Following paper nomenclature
```

File: braindecode/models/fbcnet.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            batch_size,
+            channels,
+            self.stride_factor,
+            self.n_times_padded // self.stride_factor,
+        )
+
+        x = self.temporal_layer(x)  # type: ignore[operator]
+        x = self.flatten_layer(x)
+        x = self.final_layer(x)
+        return x
```
Comment (tomMoral) [inline braindecode/models/fbcnet.py:213]: Add the shapes in comments as a way to have a sanity check

File: braindecode/models/modules.py [hunk 10]
```diff
@@ -853,3 +892,135 @@ def _apply_iir(x: Tensor, filter: dict) -> Tensor:
...
+        self.min_var = min_var
+        self.max_var = max_var
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        var = x.var(dim=self.dim, keepdim=self.keepdim)
+        var_clamped = torch.clamp(var, min=self.min_var, max=self.max_var)
+        return torch.log(var_clamped)
+
+
+class MeanLayer(nn.Module):
```
Comment (tomMoral) [inline braindecode/models/modules.py:982]: nitpick... Move this before complex ones.

File: braindecode/models/modules.py [hunk 11]
```diff
@@ -853,3 +892,135 @@ def _apply_iir(x: Tensor, filter: dict) -> Tensor:
...
+
+    def __init__(self, dim: int):
+        super().__init__()
+        self.dim = dim
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return x.mean(dim=self.dim, keepdim=True)
+
+
+class MaxLayer(nn.Module):
```
Comment (tomMoral) [inline braindecode/models/modules.py:1005]: ditto

File: braindecode/models/modules.py [hunk 12]
```diff
@@ -853,3 +892,135 @@ def _apply_iir(x: Tensor, filter: dict) -> Tensor:
...
+    def __init__(
+        self,
+        dim: int,
+        keepdim: bool = True,
+        min_var: float = 1e-6,
+        max_var: float = 1e6,
+    ):
+        super().__init__()
+        self.dim = dim
+        self.keepdim = keepdim
```
Comment (tomMoral) [inline braindecode/models/modules.py:972]: If you put keep dim for one, you should do it for all the others.

File: test/unit_tests/models/test_models.py [hunk 13]
```diff
@@ -1094,3 +1095,136 @@ def test_parameters_EEGTCNet():
...
+def test_fbcnet_invalid_temporal_layer():
+    with pytest.raises(NotImplementedError):
+        FBCNet(
+            n_chans=22,
+            n_outputs=2,
+            n_times=1000,
+            temporal_layer='InvalidLayer',
+            sfreq=250,
+        )
+
```
Comment (tomMoral) [inline test/unit_tests/models/test_models.py:1213]: Move this to the bottom

**Review by tomMoral — COMMENTED** [review]
A first pass on the review.

Improving the results to better match the one from the original paper is necessary, and checking that similar result is obtain with external filterbank will also help.

**Comment by bruAristimunha** [comment]
The to-do list here is a little more challenging:

1) match the results or discover the reason of why eegnetv4 is so low.

The same effect is observed with deep4net.


2) compare the filter bank as layer vs as  pre-processing step.

**Comment by tomMoral** [comment]
TODOs after IRL meeting:
- [ ]  do a `braindecode` example to compare an architecture with `FBLayer` and with a preprocessing to show this gives similar results with the `FBCNet`  (so simply changing the forward in the architecture in the example.
- [ ] Fix EEGNet and make sure we get around 70ish acuracy
- [ ] Train this network for longer and check we reach nice accuracy too (above 70).

File: braindecode/models/modules.py [hunk 14]
```diff
@@ -354,6 +354,76 @@ def _max_norm(self):
...
+        Classical parameters for the nn.Conv2d Layer.
+    max_norm : float, default=1.0
+        Norm for weight normalization.
+
+    Returns
+    -------
+    torch.Tensor
+        Output tensor after applying the convolutional layer.
+    """
+    def __init__(self, *args, **kwargs, max_norm=1):
```
Comment (tomMoral) [inline braindecode/models/modules.py:412]:
```suggestion
    def __init__(self, *args, max_norm=1, **kwargs):
```

File: braindecode/models/modules.py [hunk 15]
```diff
@@ -354,6 +354,76 @@ def _max_norm(self):
...
+            )
+
+
+class Conv2dWithConstraint(nn.Conv2d):
+    """
+    Convolutional layer with weight normalization constraint.
+
+    Parameters
+    ----------
+    *args, **kwargs : list, dict
```
Comment (tomMoral) [inline braindecode/models/modules.py:402]:
```suggestion
    *args, **kwargs : list, dict
```

File: braindecode/models/modules.py [hunk 16]
```diff
@@ -354,6 +354,76 @@ def _max_norm(self):
...
+        Output tensor after applying the convolutional layer.
+    """
+    def __init__(self, *args, max_norm=1, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.max_norm = max_norm
+
+    def forward(self, x):
+        self._constraint()
+        return super().forward(x)
+
```
Comment (tomMoral) [inline braindecode/models/modules.py:419]:
```suggestion
        return super().forward(x)

```

File: braindecode/models/modules.py [hunk 17]
```diff
@@ -354,6 +354,76 @@ def _max_norm(self):
...
+    *args, **kwargs : list, dict 
+        Classical parameters for the nn.Conv2d Layer.
+    max_norm : float, default=1.0
+        Norm for weight normalization.
+
+    Returns
+    -------
+    torch.Tensor
+        Output tensor after applying the convolutional layer.
+    """
```
Comment (tomMoral) [inline braindecode/models/modules.py:411]:
```suggestion
    """

```

**Comment by bruAristimunha** [comment]
I will start again @tomMoral, too complicate merge...
